### ISSUE TITLE
RFC DeviceQueues

### ISSUE THREAD

**Issue by knappador** [vulkano-rs/vulkano#1189]
Proposing a new API.  The [type graph of Vulkan objects](https://gpuopen.com/understanding-vulkan-objects/) indicates that the Queue's and Devices are quite isolated in terms of API reachability.

It's very common to only use one logical device for full API flexibility between the queue families.  Transferring data across queue familieis is only possible if the underlying `vkDevice` is identical.  Vulkan docs state this explicitly.  There is no reason as far as I can tell to even expose such Queue construction paths publicly to users.

It's unknown up front how many queues the rest of the program will want to use.  It is also unknown if the application chooses or chooses not to utilize the queues in a thread-safe manner.  The API for running commands usually involves cooperation via semaphores etc and not by coordinating the exact timing of command submission, so I think it's appropriate to actually hide whether a Queue might block on a RwLock for command buffer submission.

```
enum QueueCapability {
    Compute,
    Transfer,
    Sparse,
    Graphics,
}

impl QueueCapability {
    pub fn dedicated_compute () -> Vec<Vec<QueueCapability>>);
    pub fn universal () -> Vec<Vec<QueueCapability>>; 
    pub fn dedicated_transfer () -> Vec<Vec<QueueCapability>>;
    pub fn dedicated_sparse () -> Vec<Vec<QueueCapability>>;
    // .. other common capability orders of preference
}

impl DeviceQueues {
    /// These queues may be exhausted if you attempt to create too many
    pub fn private_queue (&self, IntoIter<QueueType>) -> Option<Queue>;
    /// These queues cannot be exhausted but implement a read-write lock
    pub fn shared_queue (&self, IntoIter<QueueType>) -> Option<Queue>;
}
```

In this implementation, I want to deviate from the pattern of tossing around Arc's and instead properly implement Send, Sync, Clone, etc for the relevant objects.  The Queue's themselves will likely have an unchanged API.

For some code to look at:
https://github.com/e-nguyen/e-nguyen/blob/master/src/ewin.rs

As you can see, I've got a `SwapWindow` prototype built up that makes it very quick to get from nothing to something useful for drawing on.  My Queue management turned out to be incomplete to do my compute in a compute queue and I wound up hacking some ugly stuff in.

**Comment by mitchmindtree** [comment]
@knappador just to clarify - are you hoping to add a new `DeviceQueues` type that allows you to specify a set of queue capabilities that you require and retrieve a `Queue` from the device that satisfies those requirements? Are you hoping to remove or change some other parts of the API in favour of this? If so, what exactly are you hoping to remove?

>  There is no reason as far as I can tell to even expose some of the Queue API's publicly to users.

What APIs exactly do you think should not be exposed?

> As you can see, I've got a SwapWindow prototype built up that makes it very quick to get from nothing to something useful for drawing on. My Queue management turned out to be incomplete to do my compute in a compute queue and I wound up hacking some ugly stuff in.

I can see that you have some high-level wrapping here to create one window and start working with it quickly, but how exactly is this related to the rest of your RFC? This is a large amount of code to sift through and I don't really understand what I'm looking for in it.

----

Currently I'm pretty happy with the existing API around queues and its flexibility. I have also built some large downstream abstractions around vulkano's devices and queues that makes it easier to work with them in a correct manner for my use cases (which often involve many-window setups for installations and sharing GPU compute across multiple physical devices and in turn different device queues). This RFC slightly concerns me as it sounds like you're suggesting some pretty sweeping breaking changes to what is a pretty core API to vulkano. Maybe some clarification on to some of my Qs above will help though!

**Comment by knappador** [comment]
> Are you hoping to remove or change some other parts of the API in favour of this? If so, what exactly are you hoping to remove?

Additive

> What APIs exactly do you think should not be exposed?

Instantiating multiple logical devices of the same physical device and winding up with two sets of queues to the same physical device that cannot for instance have a memory barrier created between them to transfer ownership.

Inverting this relationship, a wrapper that only exposes the subset of construction paths where all queues use the same `vkDevice` doesn't lose or preclude functionality and only reduces boilerplate.

In cases where multiple `vkDevice` instances for a physical device are desired, besides relying on the existing lower level API, it would be simply a matter of creating a new `DeviceQueues` instance and proceeding normally.

**Comment by knappador** [comment]
I believe an application in fact should know which queues it wants to use up front and which queues it can tolerate vs which queues it positively must have.  This should be reflected up front in the API to create a logical device wrapper.

(The same can be said of the device capabilities / extensions but this creeps scope)

I'll propose an updated API later.

**Comment by knappador** [comment]
I'm soliciting opinions and feedback on queue dispatch heuristics.  Hueristics has to be stressed because no clients are going to generate optimal queue utilization for downstream app components without advance knowledge of the target hardware and how each app component should be prioritized. (No user would be precluded from doing this, but they wouldn't use the high-level API being developed).

It's important to consider what gains should be achievable.  Slide 26 shows one such mechanism of gains. 
https://www.khronos.org/assets/uploads/developers/library/2016-vulkan-devday-uk/9-Asynchonous-compute.pdf

If an implementation uses an abstract vocabulary for downstream to request resources, it might include something like:

```rust
enum QueueRoles {
    TransferUpload,
    TransferDownload,
    Graphics,
    Compute,
    Sparse,
}
```
Using the transfer upload / download independence as an example, if both `TransferUpload` and `TransferDownload` are being requested, we would like to see the following outcome for hardware capabilities:

- ***dedicated transfer family with two queues*** - use one for upload and one for download
- ***dedicated transfer with one queue*** - use the same one for upload and download (and assume the driver does something smart)
- ***no dedicated transfer queues*** - use graphics or transfer queues depending on which ones are more over-subscribed.  Try not to use a queue that is subscribed in another role.

Examples of more general heuristics to give downstream a best-effort slice of queues:

- Try to hand out a queue with the more distinct capability;  graphics for graphics, compute for compute etc.
- Try to ensure that low priority work isn't directed onto a high priority queue
- Allow some mechanism to specify that a group of queue requests should receive independent queues and avoid subscription of a queue to more than one request in that group
- Try to give at least one exclusive queue to the highest priority and only over-subscribe queues for work below a certain priority level

I believe this makes the struct for an abstract queue request into something like:
```rust
struct QueueRequest {
    unique_group: i32,
    queue_role: QueueRole,
    priority: f32,
    can_display: bool,  // futureproof API for platform-specific WSI detection
}
```
**Handling queue over-subscription, the re-purposing that will occur when there are not enough queues to just naively hand out independent queues for each request is the most important behavior to consider.**  Where things get difficult is closer to the minimum specification corner case, where everything is destined to be over-subscribed and maybe the best we can do is honor specified independence for `unique_group` if n =2 is the largest unique group.

There's a lot of ways to try to prioritize how queues get doled out.  The big questions are:

- **What are known terrible choices**
- **What's generally desirable about a good dispatch scheme**
- **How can we define an abstract requirement that makes the dispatch logic reasonable to implement**

*E-Nguyen's implementation is being actively written.  Any comments will inform that design and perhaps yield something nice for Vulkano.  It's a design for worst-case both in terms of expected hardware range and variable application demands that are not limited to just the queue problem.*

**Comment by knappador** [comment]
The TL:DR of feedback I'm getting on other channels and just more reading says to divide work onto separate queues & command buffers if the work **both** waits on different memory dependencies and saturates on different GPU resources.  Failure to achieve either of these just means the same bubbles just with more queues.

As for writing the API to steer users towards these behaviors, I don't think `unique_group` is either good at encouraging users to understand _why_ their code is sub-optimal nor informative on the heuristic side for avoiding placing ALU-heavy work with other ALU-heavy work etc.

Since "ALU-intensive" etc are all relative to the specific hardware, truly getting mileage out of a complex queue assignment heuristic might involve packaging benchmarking software to calibrate per device.

Still, if we communicate the concerns in the API, we are both future-proof and encouraging the right usage behavior.

I want to recommend something like this making it into the API:
```rust
enum WorkloadWeight {
    ComputeHeavy,
    SamplerHeavy,
    MemoryHeavy,
    Uploads,
    Downloads,
    ...
} 
```

I need to acquire more knowledge of which resources can independently saturate.  If we have the full enumeration in the API, I believe we would be pretty much able to support most use cases for splitting work across queues if not yet capable of informing downstream of how to tune its use of resources.

**Comment by knappador** [comment]
Closing.  I don't think this thread is able to progress coherently.

The original problem statement was around preventing the queue & device relationship from so easily permitting senseless usages such as combining calls with queues from differently constructed logical devices.  

I believed that API graph reduction into a single construction could be a solution.  However, using a construction approach quickly turns into a sprawling dependency management problem as attempting to constrain the creation of queues from a device without sacrificing control becomes a study on decidable reductions in optimal queue dispatches.

Too much of this information is way outside of intended scope.  :bomb: :boom: :do_not_litter: :dancer:
